### PR TITLE
Add three modes for prelu_op

### PR DESCRIPTION
**Pull request by jerrywgz**
Three modes for prelu_op are created for weight sharing. This op could be used for CPU.

Fix https://github.com/PaddlePaddle/Paddle/issues/12350

### CODE REVIEW

**Review by wanghaoshuang — DISMISSED** [review]
LGTM.

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -5354,3 +5355,35 @@ def rank_loss(label, left, right, name=None):
                 "Right": right},
         outputs={'Out': out})
     return out
+
+
+def prelu(x, mode):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]:
```
def prelu(x, mode, name=None)
```

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -5354,3 +5355,35 @@ def rank_loss(label, left, right, name=None):
...
+    """
+    Equation:
+
+        f(x) =
+        \begin{cases}
+        \alpha * x, \quad  \text{if} \ x < 0 \\
+        x,         \qquad  \text{if} \ x >= 0
+        \end{cases}
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: 公式格式有误，需要参考其他公式写法。

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -5354,3 +5355,35 @@ def rank_loss(label, left, right, name=None):
...
+
+    Args:
+        x (Variable): The input tensor.
+        mode (string): The mode for weight sharing
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: 用户看到这里的文档，不清楚mode都可以设置啥。

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -5354,3 +5355,35 @@ def rank_loss(label, left, right, name=None):
...
+
+    Returns:
+        Variable: The output tensor with the same shape as input.
+
+    Examples:
+
+        .. code-block:: python
+
+            output = fluid.layers.prelu(x,mode)
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: 示例的mode给个明确的。

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -5354,3 +5355,35 @@ def rank_loss(label, left, right, name=None):
...
+    """
+    helper = LayerHelper('relu', **locals())
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: relu -> prelu

File: python/paddle/fluid/layers/nn.py [hunk 6]
```diff
@@ -5361,3 +5362,49 @@ def rank_loss(label, left, right, name=None):
                 "Right": right},
         outputs={'Out': out})
     return out
+
+
+def prelu(x, param_attr=None, mode, name=None):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: also need unit test in https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/tests/unittests/test_layers.py

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -5361,3 +5362,49 @@ def rank_loss(label, left, right, name=None):
...
+
+    Examples:
+
+        .. code-block:: python
+	    x = fluid.layers.data(name="x", shape=[10,10], dtype="float32")
+	    mode = 'channel'
+            output = fluid.layers.prelu(x,mode)
+    """
+    helper = LayerHelper('prelu', **locals())
+    alpha_shape = [1]
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]:
```python
if mode not in ['all', 'channel', 'element']:
    raise ValueError('mode should be one of ['all', 'channel', 'element'].')
```

File: python/paddle/fluid/layers/nn.py [hunk 8]
```diff
@@ -5361,3 +5362,49 @@ def rank_loss(label, left, right, name=None):
...
+    Args:
+        x (Variable): The input tensor.
+	param_attr(ParamAttr|None): The parameter attribute for the learnable
+                                    weight (alpha).
+        mode (string): The mode for weight sharing
+		       all: all elements share same weight
+ 		       channel:elements in a channel share same weight
+ 		       element:each element has a weight
+	name(str|None): A name for this layer(optional). If set None, the layer
+                        will be named automatically.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: Args下面的注意对齐~

File: python/paddle/fluid/layers/nn.py (hunk 1 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/nn.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/nn.py]: 已修改

File: python/paddle/fluid/tests/unittests/test_prelu_op.py [hunk 9]
```diff
@@ -20,30 +20,62 @@
...
-        x_np = np.random.normal(size=(10, 10)).astype("float32")
-
-        for pos, val in np.ndenumerate(x_np):
-            # Since zero point in prelu is not differentiable, avoid randomize
-            # zero.
-            while abs(val) < 1e-3:
-                x_np[pos] = np.random.normal()
-                val = x_np[pos]
+        self.initTestCase()
+        x_np = np.random.normal(size=(2, 2, 2)).astype("float32")
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_prelu_op.py]: 这里的size可以适当的调大点吧~

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/nn.py]: 已设置

File: python/paddle/fluid/layers/nn.py (hunk 4 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/nn.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 5 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/nn.py]: Done

File: python/paddle/fluid/tests/unittests/test_prelu_op.py [hunk 10]
```diff
@@ -20,30 +20,62 @@
...
+        if self.attrs == {'mode': "all"}:
+            alpha_np = np.array([.1], dtype="float32")
+            self.inputs = {'X': x_np, 'Alpha': alpha_np}
+        elif self.attrs == {'mode': "channel"}:
+            alpha_np = np.array([.1, .2], dtype="float32").reshape([1, 2, 1])
+            self.inputs = {'X': x_np, 'Alpha': alpha_np}
+        else:
+            alpha_np = np.array(
+                [.1, .2, .1, .2, .1, .2, .1, .2],
+                dtype="float32").reshape([2, 2, 2])
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_prelu_op.py]: alpha_np可以 np.random生成吧~

File: python/paddle/fluid/tests/unittests/test_prelu_op.py (hunk 9 above)
Comment (jerrywgz) [reply python/paddle/fluid/tests/unittests/test_prelu_op.py]: 这里调大size发现单测会有错

File: python/paddle/fluid/layers/nn.py (hunk 8 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/nn.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 6 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/nn.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 7 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/nn.py]: Done

**Review by qingqing01 — APPROVED** [review]

**Review by panyx0718 — APPROVED** [review]

File: python/paddle/fluid/tests/unittests/test_prelu_op.py [hunk 11]
```diff
@@ -20,30 +20,58 @@
...
+            alpha_np = np.random.rand(*x_np.shape).astype("float32")
+            self.inputs = {'X': x_np, 'Alpha': alpha_np}
+
         out_np = np.maximum(self.inputs['X'], 0.)
         out_np = out_np + np.minimum(self.inputs['X'],
                                      0.) * self.inputs['Alpha']
         assert out_np is not self.inputs['X']
         self.outputs = {'Out': out_np}
 
+    def initTestCase(self):
```
Comment (velconia) [inline python/paddle/fluid/tests/unittests/test_prelu_op.py:46]: Is the order of execution matters? I meet an order problem in ```Python``` @jerrywgz

**Pull request merged**